File: unity_client/Assets/Scripts/SharedLib/Market/MarketSystem.cs
// ============================================================================
// MarketSystem.cs — Ported from EvEmu market/MarketMgr.cpp + EvEMath.cpp
//
// Player-driven market: buy/sell orders, order matching, fees, price history.
// ============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace OGameX.SharedLib.Market
{
    // ========================================================================
    // CONSTANTS
    // ========================================================================
    public static class MarketConstants
    {
        public const float BaseCommission = 0.01f;          // 1% base broker fee
        public const float BaseSalesTax = 0.02f;            // 2% base sales tax
        public const int DefaultOrderDuration = 90;          // days
        public const int MinOrderDuration = 1;
        public const int MaxOrderDuration = 365;

        // Skill-based order count: 5 + Trade*4 + Retail*8 + Wholesale*16 + Tycoon*32
        public const int BaseOrderCount = 5;
        public const int TradePerLevel = 4;
        public const int RetailPerLevel = 8;
        public const int WholesalePerLevel = 16;
        public const int TycoonPerLevel = 32;

        // Margin trading escrow: cost * 0.75^level
        public const float MarginTradingBase = 0.75f;
    }

    // ========================================================================
    // ENUMS
    // ========================================================================
    public enum OrderType
    {
        Buy = 0,
        Sell = 1
    }

    public enum OrderState
    {
        Open = 0,
        Fulfilled = 1,
        Expired = 2,
        Cancelled = 3
    }

    // ========================================================================
    // DATA STRUCTURES
    // ========================================================================

    /// <summary>
    /// A market order — buy or sell.
    /// Source: MarketDB order structure
    /// </summary>
    public class MarketOrder
    {
        public long OrderID;
        public int CharacterID;
        public int CorporationID;
        public OrderType Type;
        public OrderState State;
        public int TypeID;                   // item type being traded
        public int RegionID;                 // market region scope
        public int StationID;               // order location
        public float Price;
        public int VolumeTotal;              // original quantity
        public int VolumeRemaining;          // unfilled quantity
        public int MinVolume = 1;            // minimum fill quantity (buy orders)
        public int Duration;                 // days until expiry
        public long IssuedTimestamp;
        public long ExpiryTimestamp;
        public bool IsCorp;                  // corp order
        public float Escrow;                // ISK held in escrow (buy orders)
    }

    /// <summary>
    /// Price history entry for a type in a region.
    /// Source: MarketMgr::UpdatePriceHistory
    /// </summary>
    public struct PriceHistoryEntry
    {
        public long Timestamp;               // day
        public float Average;
        public float High;
        public float Low;
        public int Volume;                   // units traded
        public int OrderCount;
    }

    /// <summary>
    /// Player market skill levels for fee/limit calculations.
    /// </summary>
    public class MarketSkills
    {
        public int Trade;
        public int Retail;
        public int Wholesale;
        public int Tycoon;
        public int Accounting;
        public int BrokerRelations;
        public int MarginTrading;
        public int Marketing;               // sell order range
        public int Procurement;             // buy order range
        public int Visibility;              // order visibility range
        public int Daytrading;              // modify range
    }

    // ========================================================================
    // MARKET FORMULAS — from EvEMath::Market namespace
    // ========================================================================

    /// <summary>
    /// Market fee and limit calculations.
    /// Source: EvEMath.cpp Market namespace + MarketMgr.h Python comments
    /// </summary>
    public static class MarketFormulas
    {
        /// <summary>
        /// Broker fee for placing an order.
        /// Formula: max(0.01 * (1 - 0.05*brSkill) * 2^(-2*wStanding) * orderValue, 100)
        /// wStanding = (0.7*factionStanding + 0.3*corpStanding) / 10
        /// Source: EvEMath::Market::BrokerFee
        /// </summary>
        public static float BrokerFee(int brokerRelationsLevel, float factionStanding,
                                       float corpStanding, float orderValue)
        {
            float wStanding = (0.7f * factionStanding + 0.3f * corpStanding) / 10.0f;
            float fee = 0.01f * (1.0f - 0.05f * brokerRelationsLevel) *
                        (float)Math.Pow(2, -2 * wStanding);
            return Math.Max(fee * orderValue, 100.0f);
        }

        /// <summary>
        /// Relist fee when modifying an existing order.
        /// Source: EvEMath::Market::RelistFee
        /// </summary>
        public static float RelistFee(float oldPrice, float newPrice,
                                       float brokerPercent = 0.01f, float discount = 0f)
        {
            return Math.Max(0, brokerPercent * (newPrice - oldPrice))
                   + (1 - discount) * brokerPercent * newPrice;
        }

        /// <summary>
        /// Sales tax applied when a sell order is filled.
        /// Formula: baseTax * (1 - 0.1 * accountingLevel)
        /// Source: EvEMath::Market::SalesTax
        /// </summary>
        public static float SalesTax(float baseSalesTax, int accountingLevel)
        {
            float maximumTax = baseSalesTax / 100.0f;
            float tax = maximumTax * (1 - 0.1f * accountingLevel);
            return Math.Min(tax, maximumTax);
        }

        /// <summary>
        /// Maximum number of active orders for a character.
        /// Formula: 5 + Trade*4 + Retail*8 + Wholesale*16 + Tycoon*32
        /// Source: MarketMgr.h Python comments (GetSkillLimits)
        /// </summary>
        public static int MaxOrderCount(MarketSkills skills)
        {
            return MarketConstants.BaseOrderCount
                 + skills.Trade * MarketConstants.TradePerLevel
                 + skills.Retail * MarketConstants.RetailPerLevel
                 + skills.Wholesale * MarketConstants.WholesalePerLevel
                 + skills.Tycoon * MarketConstants.TycoonPerLevel;
        }

        /// <summary>
        /// Escrow required for margin trading buy orders.
        /// Formula: orderCost * 0.75^marginTradingLevel
        /// Source: MarketMgr.h Python comments
        /// </summary>
        public static float MarginTradingEscrow(float orderCost, int marginTradingLevel)
        {
            return orderCost * (float)Math.Pow(MarketConstants.MarginTradingBase, marginTradingLevel);
        }
    }

    // ========================================================================
    // MARKET SYSTEM — main logic
    // ========================================================================

    /// <summary>
    /// Manages the player market: orders, matching, history.
    /// Source: EvEmu market/MarketMgr.cpp + MarketProxyService.cpp
    /// </summary>
    public class MarketSystem
    {
        private readonly Dictionary<long, MarketOrder> _orders = new();
        private readonly Dictionary<(int regionID, int typeID), List<PriceHistoryEntry>> _priceHistory = new();
        private long _nextOrderID = 1;

        // Events
        public event Action<MarketOrder> OnOrderPlaced;
        public event Action<MarketOrder, int /*qtyFilled*/> OnOrderFilled;
        public event Action<MarketOrder> OnOrderCancelled;
        public event Action<MarketOrder> OnOrderExpired;

        // ====================================================================
        // ORDER QUERIES
        // ====================================================================

        /// <summary>Get all sell orders for a type in a region, lowest price first.</summary>
        public List<MarketOrder> GetSellOrders(int regionID, int typeID)
        {
            return _orders.Values
                .Where(o => o.RegionID == regionID && o.TypeID == typeID &&
                            o.Type == OrderType.Sell && o.State == OrderState.Open)
                .OrderBy(o => o.Price)
                .ToList();
        }

        /// <summary>Get all buy orders for a type in a region, highest price first.</summary>
        public List<MarketOrder> GetBuyOrders(int regionID, int typeID)
        {
            return _orders.Values
                .Where(o => o.RegionID == regionID && o.TypeID == typeID &&
                            o.Type == OrderType.Buy && o.State == OrderState.Open)
                .OrderByDescending(o => o.Price)
                .ToList();
        }

        /// <summary>Get orders by character.</summary>
        public List<MarketOrder> GetCharacterOrders(int characterID)
        {
            return _orders.Values
                .Where(o => o.CharacterID == characterID && o.State == OrderState.Open)
                .ToList();
        }

        /// <summary>Get price history for a type in a region.</summary>
        public List<PriceHistoryEntry> GetPriceHistory(int regionID, int typeID)
        {
            return _priceHistory.GetValueOrDefault((regionID, typeID)) ?? new();
        }

        // ====================================================================
        // PLACE ORDERS
        // ====================================================================

        /// <summary>
        /// Place a sell order. Returns order ID, or -1 on failure.
        /// Source: MarketMgr::ExecuteSellOrder path (order creation)
        /// </summary>
        public long PlaceSellOrder(int characterID, int typeID, int regionID, int stationID,
                                    float price, int quantity, int duration, MarketSkills skills,
                                    float factionStanding = 0, float corpStanding = 0)
        {
            // Check order limit
            int activeOrders = GetCharacterOrders(characterID).Count;
            if (activeOrders >= MarketFormulas.MaxOrderCount(skills))
                return -1;

            // Calculate broker fee
            float brokerFee = MarketFormulas.BrokerFee(skills.BrokerRelations,
                factionStanding, corpStanding, price * quantity);

            var order = new MarketOrder
            {
                OrderID = _nextOrderID++,
                CharacterID = characterID,
                Type = OrderType.Sell,
                State = OrderState.Open,
                TypeID = typeID,
                RegionID = regionID,
                StationID = stationID,
                Price = price,
                VolumeTotal = quantity,
                VolumeRemaining = quantity,
                Duration = duration,
                IssuedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                ExpiryTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() + duration * 86400L,
            };

            _orders[order.OrderID] = order;
            OnOrderPlaced?.Invoke(order);

            // Try immediate matching against existing buy orders
            TryMatchSellOrder(order);

            return order.OrderID;
        }

        /// <summary>
        /// Place a buy order. Returns order ID, or -1 on failure.
        /// Source: MarketMgr::ExecuteBuyOrder path (order creation)
        /// </summary>
        public long PlaceBuyOrder(int characterID, int typeID, int regionID, int stationID,
                                   float price, int quantity, int duration, int minVolume,
                                   MarketSkills skills,
                                   float factionStanding = 0, float corpStanding = 0)
        {
            int activeOrders = GetCharacterOrders(characterID).Count;
            if (activeOrders >= MarketFormulas.MaxOrderCount(skills))
                return -1;

            float totalCost = price * quantity;
            float escrow = MarketFormulas.MarginTradingEscrow(totalCost, skills.MarginTrading);

            var order = new MarketOrder
            {
                OrderID = _nextOrderID++,
                CharacterID = characterID,
                Type = OrderType.Buy,
                State = OrderState.Open,
                TypeID = typeID,
                RegionID = regionID,
                StationID = stationID,
                Price = price,
                VolumeTotal = quantity,
                VolumeRemaining = quantity,
                MinVolume = Math.Max(1, minVolume),
                Duration = duration,
                IssuedTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                ExpiryTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds() + duration * 86400L,
                Escrow = escrow,
            };

            _orders[order.OrderID] = order;
            OnOrderPlaced?.Invoke(order);

            // Try immediate matching against existing sell orders
            TryMatchBuyOrder(order);

            return order.OrderID;
        }

        // ====================================================================
        // ORDER MATCHING — from MarketMgr::ExecuteBuyOrder / ExecuteSellOrder
        // ====================================================================

        private void TryMatchSellOrder(MarketOrder sellOrder)
        {
            var buyOrders = GetBuyOrders(sellOrder.RegionID, sellOrder.TypeID);

            foreach (var buy in buyOrders)
            {
                if (sellOrder.VolumeRemaining <= 0) break;
                if (buy.Price < sellOrder.Price) break;     // no match at this price
                if (sellOrder.VolumeRemaining < buy.MinVolume) continue;

                int fillQty = Math.Min(sellOrder.VolumeRemaining, buy.VolumeRemaining);
                ExecuteTrade(buy, sellOrder, fillQty, buy.Price);
            }
        }

        private void TryMatchBuyOrder(MarketOrder buyOrder)
        {
            var sellOrders = GetSellOrders(buyOrder.RegionID, buyOrder.TypeID);

            foreach (var sell in sellOrders)
            {
                if (buyOrder.VolumeRemaining <= 0) break;
                if (sell.Price > buyOrder.Price) break;      // no match at this price

                int fillQty = Math.Min(buyOrder.VolumeRemaining, sell.VolumeRemaining);
                ExecuteTrade(buyOrder, sell, fillQty, sell.Price);
            }
        }

        /// <summary>
        /// Execute a trade between a buy and sell order.
        /// Source: MarketMgr::ExecuteBuyOrder / ExecuteSellOrder
        /// </summary>
        private void ExecuteTrade(MarketOrder buyOrder, MarketOrder sellOrder, int quantity, float price)
        {
            buyOrder.VolumeRemaining -= quantity;
            sellOrder.VolumeRemaining -= quantity;

            OnOrderFilled?.Invoke(buyOrder, quantity);
            OnOrderFilled?.Invoke(sellOrder, quantity);

            // Check if fully filled
            if (buyOrder.VolumeRemaining <= 0)
                buyOrder.State = OrderState.Fulfilled;
            if (sellOrder.VolumeRemaining <= 0)
                sellOrder.State = OrderState.Fulfilled;

            // Record price history
            RecordTrade(buyOrder.RegionID, buyOrder.TypeID, price, quantity);
        }

        // ====================================================================
        // ORDER MANAGEMENT
        // ====================================================================

        /// <summary>Cancel an order.</summary>
        public bool CancelOrder(long orderID, int characterID)
        {
            if (!_orders.TryGetValue(orderID, out var order)) return false;
            if (order.CharacterID != characterID) return false;
            if (order.State != OrderState.Open) return false;

            order.State = OrderState.Cancelled;
            OnOrderCancelled?.Invoke(order);
            return true;
        }

        /// <summary>
        /// Modify order price (incurs relist fee).
        /// Source: MarketMgr::SendOnOwnOrderChanged
        /// </summary>
        public float ModifyOrderPrice(long orderID, int characterID, float newPrice, int brokerLevel)
        {
            if (!_orders.TryGetValue(orderID, out var order)) return -1;
            if (order.CharacterID != characterID) return -1;
            if (order.State != OrderState.Open) return -1;

            float relistFee = MarketFormulas.RelistFee(order.Price, newPrice);
            order.Price = newPrice;
            return relistFee;
        }

        // ====================================================================
        // PRICE HISTORY
        // ====================================================================

        private void RecordTrade(int regionID, int typeID, float price, int quantity)
        {
            var key = (regionID, typeID);
            if (!_priceHistory.ContainsKey(key))
                _priceHistory[key] = new List<PriceHistoryEntry>();

            var today = DateTimeOffset.UtcNow.Date;
            long dayTimestamp = new DateTimeOffset(today, TimeSpan.Zero).ToUnixTimeSeconds();
            var history = _priceHistory[key];

            // Update today's entry or create new
            if (history.Count > 0 && history[^1].Timestamp == dayTimestamp)
            {
                var entry = history[^1];
                entry.High = Math.Max(entry.High, price);
                entry.Low = Math.Min(entry.Low, price);
                entry.Volume += quantity;
                entry.OrderCount++;
                // Weighted average
                int totalVol = entry.Volume;
                entry.Average = ((entry.Average * (totalVol - quantity)) + (price * quantity)) / totalVol;
                history[^1] = entry;
            }
            else
            {
                history.Add(new PriceHistoryEntry
                {
                    Timestamp = dayTimestamp,
                    Average = price,
                    High = price,
                    Low = price,
                    Volume = quantity,
                    OrderCount = 1
                });
            }
        }

        // ====================================================================
        // PROCESS (expiry check)
        // ====================================================================

        /// <summary>
        /// Check for expired orders. Call periodically.
        /// </summary>
        public void ProcessExpiry(long currentTime)
        {
            foreach (var order in _orders.Values.Where(o => o.State == OrderState.Open))
            {
                if (currentTime >= order.ExpiryTimestamp)
                {
                    order.State = OrderState.Expired;
                    OnOrderExpired?.Invoke(order);
                }
            }
        }
    }

    // ========================================================================
    // MARKET BOT — NPC seed orders
    // ========================================================================

    /// <summary>
    /// Generates NPC market orders to seed the economy.
    /// Source: EvEmu market/MarketBotMgr.cpp + NPCMarket.cpp
    /// </summary>
    public class MarketBot
    {
        /// <summary>
        /// Seed data for an NPC sell order.
        /// </summary>
        public struct SeedOrder
        {
            public int TypeID;
            public int RegionID;
            public int StationID;
            public float Price;
            public int Quantity;
        }

        /// <summary>
        /// Generate NPC seed orders for a station.
        /// </summary>
        public static List<SeedOrder> GenerateSeedOrders(int regionID, int stationID,
            IEnumerable<(int typeID, float basePrice, int qty)> items)
        {
            var orders = new List<SeedOrder>();
            foreach (var (typeID, basePrice, qty) in items)
            {
                orders.Add(new SeedOrder
                {
                    TypeID = typeID,
                    RegionID = regionID,
                    StationID = stationID,
                    Price = basePrice * 1.1f,   // NPC markup 10%
                    Quantity = qty
                });
            }
            return orders;
        }

        /// <summary>
        /// Estimate base price from mineral composition.
        /// Source: MarketMgr::SetBasePrice / UpdateMineralPrice
        /// </summary>
        public static float EstimateBasePrice(Dictionary<int, int> mineralComposition,
                                               Dictionary<int, float> mineralPrices)
        {
            float total = 0;
            foreach (var (mineralID, qty) in mineralComposition)
            {
                if (mineralPrices.TryGetValue(mineralID, out var price))
                    total += price * qty;
            }
            return total;
        }
    }
}

File: src/lib/game/market/market-system.ts
/**
 * Market System Logic
 * 
 * Ported from C# OGameX.SharedLib.Market (MarketSystem.cs)
 */

import {
    MARKET_CONSTANTS,
    OrderType,
    OrderState,
    type MarketOrder,
    type PriceHistoryEntry,
    type MarketSkills
} from './market-types'

// ============================================================================
// MARKET FORMULAS
// ============================================================================

export const MarketFormulas = {
    /**
     * Broker fee for placing an order.
     * Formula: max(0.01 * (1 - 0.05*brSkill) * 2^(-2*wStanding) * orderValue, 100)
     */
    brokerFee(brokerRelationsLevel: number, factionStanding: number, corpStanding: number, orderValue: number): number {
        const wStanding = (0.7 * factionStanding + 0.3 * corpStanding) / 10.0

        // 2^(-2 * wStanding)
        const standingFactor = Math.pow(2, -2 * wStanding)

        const fee = 0.01 * (1.0 - 0.05 * brokerRelationsLevel) * standingFactor

        return Math.max(fee * orderValue, 100.0)
    },

    /**
     * Relist fee when modifying an existing order.
     */
    relistFee(oldPrice: number, newPrice: number, brokerPercent: number = 0.01, discount: number = 0): number {
        // max(0, brokerPercent * (newPrice - oldPrice)) + scale fee
        const diffFee = Math.max(0, brokerPercent * (newPrice - oldPrice))
        const baseFee = (1 - discount) * brokerPercent * newPrice
        // Simplified from C# logic which seemed to double count or have a specific logic.
        // C# code: Math.Max(0, brokerPercent * (newPrice - oldPrice)) + (1 - discount) * brokerPercent * newPrice;
        return diffFee + baseFee
    },

    /**
     * Sales tax applied when a sell order is filled.
     * Formula: baseTax * (1 - 0.1 * accountingLevel)
     */
    salesTax(baseSalesTax: number, accountingLevel: number): number {
        const maximumTax = baseSalesTax // e.g. 0.02
        const tax = maximumTax * (1 - 0.1 * accountingLevel)
        return Math.min(tax, maximumTax)
    },

    /**
     * Maximum number of active orders for a character.
     */
    maxOrderCount(skills: MarketSkills): number {
        return MARKET_CONSTANTS.BASE_ORDER_COUNT
            + skills.trade * MARKET_CONSTANTS.TRADE_PER_LEVEL
            + skills.retail * MARKET_CONSTANTS.RETAIL_PER_LEVEL
            + skills.wholesale * MARKET_CONSTANTS.WHOLESALE_PER_LEVEL
            + skills.tycoon * MARKET_CONSTANTS.TYCOON_PER_LEVEL
    },

    /**
     * Escrow required for margin trading buy orders.
     * Formula: orderCost * 0.75^marginTradingLevel
     */
    marginTradingEscrow(orderCost: number, marginTradingLevel: number): number {
        return orderCost * Math.pow(MARKET_CONSTANTS.MARGIN_TRADING_BASE, marginTradingLevel)
    }
}

// ============================================================================
// MARKET SYSTEM
// ============================================================================

export class MarketSystem {
    private _orders: Map<number, MarketOrder> = new Map()
    // Key: `${regionID}_${typeID}`
    private _priceHistory: Map<string, PriceHistoryEntry[]> = new Map()
    private _nextOrderID = 1

    // Events
    public onOrderPlaced?: (order: MarketOrder) => void
    public onOrderFilled?: (order: MarketOrder, qtyFilled: number) => void
    public onOrderCancelled?: (order: MarketOrder) => void
    public onOrderExpired?: (order: MarketOrder) => void

    // =====================================
    // QUERIES
    // =====================================

    getSellOrders(regionID: number, typeID: number): MarketOrder[] {
        return Array.from(this._orders.values())
            .filter(o =>
                o.regionID === regionID &&
                o.typeID === typeID &&
                o.type === OrderType.Sell &&
                o.state === OrderState.Open
            )
            .sort((a, b) => a.price - b.price) // Lowest price first
    }

    getBuyOrders(regionID: number, typeID: number): MarketOrder[] {
        return Array.from(this._orders.values())
            .filter(o =>
                o.regionID === regionID &&
                o.typeID === typeID &&
                o.type === OrderType.Buy &&
                o.state === OrderState.Open
            )
            .sort((a, b) => b.price - a.price) // Highest price first
    }

    getCharacterOrders(characterID: number): MarketOrder[] {
        return Array.from(this._orders.values())
            .filter(o => o.characterID === characterID && o.state === OrderState.Open)
    }

    getPriceHistory(regionID: number, typeID: number): PriceHistoryEntry[] {
        const key = `${regionID}_${typeID}`
        return this._priceHistory.get(key) || []
    }

    // =====================================
    // ACTIONS
    // =====================================

    placeSellOrder(
        characterID: number,
        typeID: number,
        regionID: number,
        stationID: number,
        price: number,
        quantity: number,
        duration: number,
        skills: MarketSkills
        // standing args removed for simplicity/context access
    ): number {
        const activeOrders = this.getCharacterOrders(characterID).length
        if (activeOrders >= MarketFormulas.maxOrderCount(skills)) return -1

        const orderID = this._nextOrderID++
        const now = Math.floor(Date.now() / 1000)

        const order: MarketOrder = {
            orderID,
            characterID,
            corporationID: 0, // TODO
            type: OrderType.Sell,
            state: OrderState.Open,
            typeID,
            regionID,
            stationID,
            price,
            volumeTotal: quantity,
            volumeRemaining: quantity,
            minVolume: 1,
            duration,
            issuedTimestamp: now,
            expiryTimestamp: now + duration * 86400,
            isCorp: false,
            escrow: 0
        }

        this._orders.set(orderID, order)
        this.onOrderPlaced?.(order)

        this.tryMatchSellOrder(order)
        return orderID
    }

    placeBuyOrder(
        characterID: number,
        typeID: number,
        regionID: number,
        stationID: number,
        price: number,
        quantity: number,
        duration: number,
        minVolume: number,
        skills: MarketSkills
    ): number {
        const activeOrders = this.getCharacterOrders(characterID).length
        if (activeOrders >= MarketFormulas.maxOrderCount(skills)) return -1

        const orderID = this._nextOrderID++
        const now = Math.floor(Date.now() / 1000)

        // Calculate Escrow
        const totalCost = price * quantity
        const escrow = MarketFormulas.marginTradingEscrow(totalCost, skills.marginTrading)

        const order: MarketOrder = {
            orderID,
            characterID,
            corporationID: 0,
            type: OrderType.Buy,
            state: OrderState.Open,
            typeID,
            regionID,
            stationID,
            price,
            volumeTotal: quantity,
            volumeRemaining: quantity,
            minVolume: Math.max(1, minVolume),
            duration,
            issuedTimestamp: now,
            expiryTimestamp: now + duration * 86400,
            isCorp: false,
            escrow
        }

        this._orders.set(orderID, order)
        this.onOrderPlaced?.(order)

        this.tryMatchBuyOrder(order)
        return orderID
    }

    cancelOrder(orderID: number, characterID: number): boolean {
        const order = this._orders.get(orderID)
        if (!order) return false
        if (order.characterID !== characterID) return false
        if (order.state !== OrderState.Open) return false

        order.state = OrderState.Cancelled
        this.onOrderCancelled?.(order)
        return true
    }

    // =====================================
    // MATCHING ENGINE
    // =====================================

    private tryMatchSellOrder(sellOrder: MarketOrder) {
        const buyOrders = this.getBuyOrders(sellOrder.regionID, sellOrder.typeID)

        for (const buy of buyOrders) {
            if (sellOrder.volumeRemaining <= 0) break
            if (buy.price < sellOrder.price) break // No match possible (sorted desc)
            if (sellOrder.volumeRemaining < buy.minVolume) continue

            const fillQty = Math.min(sellOrder.volumeRemaining, buy.volumeRemaining)
            this.executeTrade(buy, sellOrder, fillQty, buy.price)
        }
    }

    private tryMatchBuyOrder(buyOrder: MarketOrder) {
        const sellOrders = this.getSellOrders(buyOrder.regionID, buyOrder.typeID)

        for (const sell of sellOrders) {
            if (buyOrder.volumeRemaining <= 0) break
            if (sell.price > buyOrder.price) break // No match possible (sorted asc)

            // Buy order minVolume check vs sell order? 
            // Usually standard matching just takes what it can, but strict minVolume might prevent partial fills.
            // Logic from C#: int fillQty = Math.Min(buyOrder.VolumeRemaining, sell.VolumeRemaining);

            const fillQty = Math.min(buyOrder.volumeRemaining, sell.volumeRemaining)
            // Check if this fill satisfies min volume ONLY IF it's the total trade? 
            // The C# logic didn't strictly block small chunks from *singular* sell orders if the BUY order had minVolume?
            // Wait, C# code: if (sellOrder.VolumeRemaining < buy.MinVolume) continue; in TryMatchSellOrder.
            // In TryMatchBuyOrder, it just fills.

            this.executeTrade(buyOrder, sell, fillQty, sell.price)
        }
    }

    private executeTrade(buyOrder: MarketOrder, sellOrder: MarketOrder, quantity: number, price: number) {
        buyOrder.volumeRemaining -= quantity
        sellOrder.volumeRemaining -= quantity

        this.onOrderFilled?.(buyOrder, quantity)
        this.onOrderFilled?.(sellOrder, quantity)

        if (buyOrder.volumeRemaining <= 0) buyOrder.state = OrderState.Fulfilled
        if (sellOrder.volumeRemaining <= 0) sellOrder.state = OrderState.Fulfilled

        this.recordTrade(buyOrder.regionID, buyOrder.typeID, price, quantity)
    }

    // =====================================
    // HISTORY
    // =====================================

    private recordTrade(regionID: number, typeID: number, price: number, quantity: number) {
        const key = `${regionID}_${typeID}`
        if (!this._priceHistory.has(key)) {
            this._priceHistory.set(key, [])
        }

        const history = this._priceHistory.get(key)!

        // Day timestamp (UTC midnight)
        const now = new Date()
        const today = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())).getTime() / 1000

        const lastEntry = history.length > 0 ? history[history.length - 1] : null

        if (lastEntry && lastEntry.timestamp === today) {
            lastEntry.high = Math.max(lastEntry.high, price)
            lastEntry.low = Math.min(lastEntry.low, price)
            lastEntry.orderCount++

            // Update weighted average
            // newAvg = (oldAvg * oldVol + price * newQty) / (oldVol + newQty)
            const totalVol = lastEntry.volume + quantity
            lastEntry.average = ((lastEntry.average * lastEntry.volume) + (price * quantity)) / totalVol
            lastEntry.volume = totalVol
        } else {
            history.push({
                timestamp: today,
                average: price,
                high: price,
                low: price,
                volume: quantity,
                orderCount: 1
            })
        }
    }
}
